### PR TITLE
Add Intsights Parser

### PR DESCRIPTION
**Pull request by 37b**
Added IntSights threat intel parser: https://intsights.com/products

### CODE REVIEW

**Comment by Maffooch** [comment]
There is a few extra commits in here that shouldn't belong. Please base against `dev` rather than `master` next time

File: dojo/tools/intsights/parser.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                              f'\r\n**Source Type**: {finding["Details"]["Source"]["Type"]}' \
+                              f'\r\n**Source Network Type**: {finding["Details"]["Source"]["NetworkType"]}' \
+                              f'\r\n\r\n----' \
+                              f'\r\n**Asset Type**: {finding["Assets"][0]["Type"]}' \
+                              f'\r\n\r\n----' \
+                              f'\r\n**Takedown Status**: {finding["TakedownStatus"]}'
+                severity = finding['Details']['Severity']
+                mitigation = "N/A"
+                impact = "N/A"
+                references = finding["Details"]["Source"]["URL"]
```
Comment (valentijnscholten) [inline dojo/tools/intsights/parser.py:55]: is this field always present?

**Comment by valentijnscholten** [comment]
can you update the title to be more descriptive? can you add some info the docs i.e. what scanner is this / where can we find more info.

File: dojo/tools/intsights/parser.py (hunk 1 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: It is present in all of the findings I have data for, but I will change it to fail safe. Will also rebase off of dev.

File: dojo/fixtures/test_type.json [hunk 2]
```diff
@@ -719,5 +719,12 @@
...
+    },
+    {
+        "fields": {
+            "name": "IntSights Report"
+        },
+        "model": "dojo.test_type",
+        "pk": 195
     }
-]
+]
```
Comment (damiencarol) [inline dojo/fixtures/test_type.json:730]: Modification of this file is useless

File: dojo/tools/intsights/parser.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        for entry in assets:
+            assets_affected = f'{assets_affected} Type: {entry["Type"]}, Value: {entry["Value"]},'
+
+        return assets_affected[:-1]
+
+    def get_findings(self, file, test):
+        duplicates = dict()
+
+        if not file:
+            return []
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:36]: This piece is useless
```suggestion
```

File: dojo/tools/intsights/parser.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+        data = file.read()
+
+        try:
+            findings = json.loads(str(data, 'utf-8'))
+        except:
+            findings = json.loads(data)
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:43]: Keep it simple please. let `json` package do the job
```suggestion
        data = json.load(file)
```

File: dojo/tools/intsights/parser.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        if not findings:
+            raise ValueError('IntSights report contains errors: No vulnerabilities were found in the data provided')
+
+        if findings.get('Findings'):
+            for finding in findings['Findings']:
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:49]:
```suggestion
        for finding in findings.get('Findings', []):
```

File: dojo/unittests/tools/test_intsights_parser.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    def get_test(self):
+        test = Test()
+        test.engagement = Engagement()
+        test.engagement.product = Product()
+        return test
+
+    def test_intsights_parser_without_file_has_no_findings(self):
+        parser = IntSightsParser()
+        findings = parser.get_findings(None, self.get_test())
+        self.assertEqual(0, len(findings))
```
Comment (damiencarol) [inline dojo/unittests/tools/test_intsights_parser.py:16]:
```suggestion
```
This use case is prohibited now. the upper layers will not call this method without a file.

File: dojo/unittests/tools/test_intsights_parser.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+        self.assertEqual("5c80dbf83b4a3900078b6be6", finding.unique_id_from_tool)
+        self.assertEqual("HTTP headers weakness in initech.com web server", finding.title)
+
+    def test_intsights_parser_with_many_vuln_has_many_findings(self):
+        testfile = open("dojo/unittests/scans/intsights/intsights_many_vul.json")
+        parser = IntSightsParser()
+        findings = parser.get_findings(testfile, self.get_test())
+        testfile.close()
+        self.assertEqual(3, len(findings))
+
```
Comment (damiencarol) [inline dojo/unittests/tools/test_intsights_parser.py:43]: Could you add more checks here?
I saw an attribute called "FoundDate": "2019-01-21T15:57:25.709Z" in reports, could be a good candidate for `finding.date`

File: dojo/unittests/tools/test_intsights_parser.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+from dojo.tools.intsights.parser import IntSightsParser
+from dojo.models import Engagement, Product, Test
+
+
+class TestIntSightsParser(TestCase):
+    def get_test(self):
+        test = Test()
+        test.engagement = Engagement()
+        test.engagement.product = Product()
+        return test
```
Comment (damiencarol) [inline dojo/unittests/tools/test_intsights_parser.py:11]: I think you can remove this piece and use directly this form in unit tests:

```
findings = parser.get_findings(file, Test())
```

File: dojo/tools/intsights/parser.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+                else:
+                    duplicates[dupe_key] = True
+
+                    finding = Finding(title = title,
+                                      test = test,
+                                      active = active,
+                                      verified = True,
+                                      description = description,
+                                      severity = severity,
+                                      numerical_severity = Finding.get_numerical_severity(severity),
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:90]:
```suggestion
```
This line is useless. it's managed outside of parsers.

File: dojo/tools/intsights/parser.py [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                              f'\r\n**Source Date**: {finding["Details"]["Source"].get("Date", "None provided")}' \
+                              f'\r\n**Source Type**: {finding["Details"]["Source"].get("URL", "N/A")}' \
+                              f'\r\n**Source Network Type**: {finding["Details"]["Source"].get("NetworkType", "N/A")}' \
+                              f'\r\n\r\n----' \
+                              f'\r\n**Assets Affected**: {assets_affected}' \
+                              f'\r\n\r\n----' \
+                              f'\r\n**Takedown Status**: {finding["TakedownStatus"]}'
+                severity = finding['Details']['Severity']
+                mitigation = "N/A"
+                impact = "N/A"
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:72]: These attribute are not mandatory for parsers.
So I think you can remove these lines and the attributes in the Finding constructor

File: docs/content/integrations/import.md [hunk 11]
```diff
@@ -320,6 +320,11 @@ Immuniweb Scan
 
 XML Scan Result File from Immuniweb Scan.
 
+IntSights Report
+--------------
+
+JSON report from IntSights ETP Suite API
+
```
Comment (damiencarol) [inline docs/content/integrations/import.md:327]: Could you add more details about where these reports come from? URL of the tools?
Specification?

**Review by damiencarol — CHANGES_REQUESTED** [review]
Made a lot of comments.

File: dojo/tools/intsights/parser.py (hunk 4 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: @damiencarol This was part of the cookie cutter boilerplate code. Should it be removed?

File: dojo/unittests/tools/test_intsights_parser.py (hunk 6 above)
Comment (37b) [reply dojo/unittests/tools/test_intsights_parser.py]: @damiencarol This is also part of the boilerplate cookie cutter code. Should it be removed?

File: dojo/unittests/tools/test_intsights_parser.py (hunk 7 above)
Comment (37b) [reply dojo/unittests/tools/test_intsights_parser.py:53]: The FoundDate is when the threat intel service found the issue. I wanted to keep that separate from when it was entered into Dojo as a Finding (affecting SLAs).

How are other issues/findings dealing with this kind of issue?

File: dojo/tools/intsights/parser.py (hunk 4 above)
Comment (damiencarol) [reply dojo/tools/intsights/parser.py:43]: yep. the cookie cutter are very new. some bloat code like this one that we used to write in the past are not welcome now. sorry for the mislead :)

File: dojo/unittests/tools/test_intsights_parser.py (hunk 8 above)
Comment (37b) [reply dojo/unittests/tools/test_intsights_parser.py]: Also boilerplate from cookie cutter template...

File: dojo/tools/intsights/parser.py (hunk 9 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: also boilerplate from cookie cutter template...

File: dojo/unittests/tools/test_intsights_parser.py (hunk 6 above)
Comment (damiencarol) [reply dojo/unittests/tools/test_intsights_parser.py:16]: yep. I see that I really need to update this boilerplate code :D

File: dojo/tools/intsights/parser.py (hunk 10 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: boilerplate from cookie cutter!

File: docs/content/integrations/import.md (hunk 11 above)
Comment (37b) [reply docs/content/integrations/import.md]: Certainly!

File: dojo/unittests/tools/test_intsights_parser.py (hunk 7 above)
Comment (damiencarol) [reply dojo/unittests/tools/test_intsights_parser.py:43]: You're right if it's the "found" date in the external DB. I was guessing it's was when the finding was detected.
Other parsers use the data of the finding or the date of the report (all finding will have the same date).
You can also take a look at other parser recently updated to see how the date is set (attribute `date`) and checked in the unit tests.

File: dojo/tools/intsights/parser.py (hunk 4 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: No, this is good. This is my first non-trivial commit so I want lots of review/info as to best practices.

**Comment by damiencarol** [comment]
@37b I pushed some PR to the cookie cutter. Hope it will help future contributions. ;)

File: dtrackvenv/include/site/python3.6/greenlet/greenlet.h [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        ((PyObject*)_PyGreenlet_API[PyExc_GreenletExit_NUM])
+
+/*
+ * PyGreenlet_New(PyObject *args)
+ *
+ * greenlet.greenlet(run, parent=None)
+ */
+#    define PyGreenlet_New                                        \
+        (*(PyGreenlet * (*)(PyObject * run, PyGreenlet * parent)) \
+             _PyGreenlet_API[PyGreenlet_New_NUM])
```
Comment (damiencarol) [inline dtrackvenv/include/site/python3.6/greenlet/greenlet.h:84]: Not sure if this file is required. :)

Comment (37b) [reply dtrackvenv/include/site/python3.6/greenlet/greenlet.h]: d'oh

**Comment by 37b** [comment]
@damiencarol Made some updates. Please provide feedback as you have time.

File: dojo/tools/intsights/parser.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+                    alerts.append(alert)
+        else:
+            self._LOGGER.error('The CSV file has one or more missing or unexpected header values')
+
+        return alerts
+
+    def get_findings(self, file, test):
+        duplicates = dict()
+
+        self._LOGGER.error(type(file))
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:120]: you could git ride of this one now I think

File: dojo/tools/intsights/parser.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+
+        if file.name.lower().endswith('.json'):
+            alerts = self._parse_json(file, )
+        elif file.name.lower().endswith('.csv'):
+            alerts = self._parse_csv(file)
+        else:
+            raise ValueError('IntSights report contains errors: Unknown File Format')
+
+        if not alerts:
+            raise ValueError('IntSights report contains errors: No vulnerabilities were found in the data provided')
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:130]: This case can't happen. I think you could remove those lines
```suggestion
```

File: dojo/tools/intsights/parser.py [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+                                              f'\r\n**Source Network Type**: {alert["network_type"]}'
+                                              f'\n\n----'
+                                              f'\r\n**Assets Affected**: {alert["assets"]}'
+                                              f'\n\n----'
+                                              f'\r\n**Alert Link**: {alert["alert_link"]}',
+                                severity=alert['severity'],
+                                references=alert["alert_link"],
+                                static_finding=False,
+                                dynamic_finding=True,
+                                unique_id_from_tool=alert['alert_id'])
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:161]: Could you move this part outside the 'if/then' ? better to manage description with duplicates

File: dojo/tools/intsights/parser.py (hunk 13 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: Thanks - was debugging with it obviously 😞

File: dojo/tools/intsights/parser.py (hunk 14 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: I have 3 tests that fail if this is removed. Basically if the file is empty, json Alerts array is empty, or the csv values are empty

File: dojo/tools/intsights/parser.py (hunk 15 above)
Comment (37b) [reply dojo/tools/intsights/parser.py]: Done!

Comment (37b) [reply dojo/tools/intsights/parser.py]: IIRC this was also from the cookie cutter template.

File: dojo/unittests/tools/test_intsights_parser.py [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    def test_intsights_parser_with_no_vuln_has_no_findings_json(self):
+        with self.assertRaises(ValueError) as context:
+            testfile = open("dojo/unittests/scans/intsights/intsights_zero_vul.json")
+            parser = IntSightsParser()
+            findings = parser.get_findings(testfile, self.get_test())
+            testfile.close()
+            self.assertTrue(
+                "IntSights report contains errors:" in str(context.exception)
+            )
+            self.assertTrue("ECONNREFUSED" in str(context.exception))
```
Comment (damiencarol) [inline dojo/unittests/tools/test_intsights_parser.py:23]: I see in this test that you want to raise an exception if the report is empty.
Is it something that you want?
Other parsers just return an empty `list`. Why do you want to raise a value error?

**Review by damiencarol — CHANGES_REQUESTED** [review]
The pull request in almost done.
There is just the use cases of empty reports that generate `ValueError` that are not aligned with orther parsers

File: dojo/tools/intsights/parser.py (hunk 14 above)
Comment (damiencarol) [reply dojo/tools/intsights/parser.py]: I see now the pb. you rely on this `ValueError` when there is no data in the report.
I prefer to do like other parsers. I mean returning an empty list.

**Review by damiencarol — APPROVED** [review]

**Comment by damiencarol** [comment]
Accepting as @37b made a lot of improvements, seems ready to review

File: dojo/tools/intsights/parser.py [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+                if alert not in invalid_alerts:
+                    alerts.append(alert)
+        else:
+            self._LOGGER.error('The CSV file has one or more missing or unexpected header values')
+
+        return alerts
+
+    def _build_finding_description(self, alert: dict) -> str:
+        """
+        Builds an IntSights FInding description from various pieces of information.
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:119]:
```suggestion
        Builds an IntSights finding description from various pieces of information.
```

File: dojo/tools/intsights/parser.py [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+                      f'\r\n**Date Found**: {alert.get("report_date", "None provided")}' \
+                      f'\n\n----' \
+                      f'\r\n**Type**: {alert.get("type", "None provided")}' \
+                      f'\r\n**Source**: {alert.get("source_url", "None provided")}' \
+                      f'\r\n**Source Date**: {alert.get("source_date", "None provided")}' \
+                      f'\r\n**Source Network Type**: {alert.get("network_type", "None provided")}' \
+                      f'\n\n----' \
+                      f'\r\n**Assets Affected**: {alert.get("assets", "None provided")}' \
+                      f'\n\n----' \
+                      f'\r\n**Alert Link**: {alert.get("alert_link", "None provided")}'
```
Comment (damiencarol) [inline dojo/tools/intsights/parser.py:136]: 1) for this kind of structure, you can use something like:
```Python
"\n".join([
  part1,
  part2,
])
```
2) as it's Markdown syntax, you can add ` chars like that:

```Python
                description = "\n".join([
                    "**IdRef:** `" + rule_result.attrib['idref'] + "`",
                    "**Title:** `" + title + "`",
                ])
```
see https://github.com/DefectDojo/django-DefectDojo/blob/4a9a4e46525874e89daa2d677679bbbe1a89494c/dojo/tools/openscap/parser.py#L54-L57

Comment (damiencarol) [reply dojo/tools/intsights/parser.py:136]: it produces better formatting when rendering the HTML
image

**Review by Maffooch — APPROVED** [review]
@damiencarol Whenever you're ready to merge

**Comment by damiencarol** [comment]
@37b @Maffooch rebased on last `dev` as there where conflicts.

**Review by damiencarol — APPROVED** [review]

**Pull request merged**